trie/trie.go

```diff
@@ -73,37 +119,62 @@ func (t *Trie[V]) Put(key string, val V) {
 	if !t.Contains(key) {
 		t.n++
 	}
-	t.root = t.put(t.root, key, val, 0, true)
+	t.root = t.put(t.root, key, val, 0)
+}
+
+func (t *Trie[V]) put(x *node[V], key string, val V, d int) *node[V] {
+	c := key[d]
+	if x == nil {
+		x = &node[V]{
+			c: c,
+		}
+	}
+	if c < x.c {
+		x.left = t.put(x.left, key, val, d)
+	} else if c > x.c {
+		x.right = t.put(x.right, key, val, d)
+	} else if d < len(key)-1 {
+		x.mid = t.put(x.mid, key, val, d+1)
+	} else {
+		x.val = val
+		x.valid = true
+	}
+	return x
 }
 
-// Remove removes the value associated with 'key'.
+// Remove removes the value associated with 'key', along with any nodes of the key that are no
+// longer used.
 func (t *Trie[V]) Remove(key string) {
-	if len(key) == 0 || !t.Contains(key) {
+	if len(key) == 0 {
 		return
 	}
-	var v V
+
+	t.root = t.remove(t.root, key, 0)
 	t.n--
-	// put a tombstone into the deleted value's node
-	t.root = t.put(t.root, key, v, 0, false)
 }
 
-func (t *Trie[V]) put(x *node[V], key string, val V, d int, valid bool) *node[V] {
-	c := key[d]
+func (t *Trie[V]) remove(x *node[V], key string, d int) *node[V] {
 	if x == nil {
-		x = &node[V]{
-			c: c,
-		}
+		return nil
 	}
+
+	c := key[d]
 	if c < x.c {
-		x.left = t.put(x.left, key, val, d, valid)
+		x.left = t.remove(x.left, key, d)
 	} else if c > x.c {
-		x.right = t.put(x.right, key, val, d, valid)
+		x.right = t.remove(x.right, key, d)
 	} else if d < len(key)-1 {
-		x.mid = t.put(x.mid, key, val, d+1, valid)
+		x.mid = t.remove(x.mid, key, d+1)
 	} else {
-		x.val = val
-		x.valid = valid
+		var v V
+		x.val = v
+		x.valid = false
```

Comment:
Is it possible for this case to happen? It seems like this still uses some tombstones if so (shouldn't we be aiming to completely remove `valid`?)

Reply:
Ah right, sorry it’s been a while since I’ve looked at this code haha.